Firestore/core/test/unit/local/leveldb_key_test.cc

```diff
@@ -490,6 +490,154 @@ TEST(NamedQueryKeyTest, Description) {
 }
 
+TEST(IndexConfigurationKeyTest, Prefixing) {
+  auto table_key = LevelDbIndexConfigurationKey::KeyPrefix();
+
+  ASSERT_TRUE(
+      absl::StartsWith(LevelDbIndexConfigurationKey::Key(0), table_key));
+
+  ASSERT_FALSE(absl::StartsWith(LevelDbIndexConfigurationKey::Key(1),
+                                LevelDbIndexConfigurationKey::Key(2)));
+}
+
+TEST(IndexConfigurationKeyTest, Ordering) {
+  ASSERT_LT(LevelDbIndexConfigurationKey::Key(0),
+            LevelDbIndexConfigurationKey::Key(1));
+  ASSERT_EQ(LevelDbIndexConfigurationKey::Key(1),
+            LevelDbIndexConfigurationKey::Key(1));
+}
+
+TEST(IndexConfigurationKeyTest, EncodeDecodeCycle) {
+  LevelDbIndexConfigurationKey key;
+
+  for (int32_t id = -5; id < 10; ++id) {
+    auto encoded = LevelDbIndexConfigurationKey::Key(id);
+    bool ok = key.Decode(encoded);
+    ASSERT_TRUE(ok);
+    ASSERT_EQ(id, key.index_id());
+  }
+}
+
+TEST(IndexConfigurationKeyTest, Description) {
+  AssertExpectedKeyDescription("[index_configuration: index_id=8]",
+                               LevelDbIndexConfigurationKey::Key(8));
+}
+
+TEST(IndexStateKeyTest, Prefixing) {
+  auto table_key = LevelDbIndexStateKey::KeyPrefix();
+
+  ASSERT_TRUE(
+      absl::StartsWith(LevelDbIndexStateKey::Key(0, "user_a"), table_key));
+
+  ASSERT_FALSE(absl::StartsWith(LevelDbIndexStateKey::Key(0, "user_a"),
+                                LevelDbIndexStateKey::Key(0, "user_b")));
+  ASSERT_FALSE(absl::StartsWith(LevelDbIndexStateKey::Key(0, "user_a"),
+                                LevelDbIndexStateKey::Key(1, "user_a")));
+}
+
+TEST(IndexStateKeyTest, Ordering) {
+  ASSERT_LT(LevelDbIndexStateKey::Key(0, "foo/bar"),
+            LevelDbIndexStateKey::Key(1, "foo/bar"));
+  ASSERT_LT(LevelDbIndexStateKey::Key(0, "foo/bar"),
+            LevelDbIndexStateKey::Key(0, "foo/bar1"));
+}
+
+TEST(IndexStateKeyTest, EncodeDecodeCycle) {
+  LevelDbIndexStateKey key;
+
+  std::vector<std::pair<int32_t, std::string>> ids{
+      {0, "foo/bar"}, {1, "foo/bar2"}, {-1, "foo-bar?baz!quux"}};
+  for (auto&& id : ids) {
+    auto encoded = LevelDbIndexStateKey::Key(id.first, id.second);
+    bool ok = key.Decode(encoded);
+    ASSERT_TRUE(ok);
+    ASSERT_EQ(id.first, key.index_id());
+    ASSERT_EQ(id.second, key.user_id());
+  }
+}
+
+TEST(IndexStateKeyTest, Description) {
+  AssertExpectedKeyDescription(
+      "[index_state: index_id=99 user_id=foo-bar?baz!quux]",
+      LevelDbIndexStateKey::Key(99, "foo-bar?baz!quux"));
+}
+
+TEST(IndexEntryKeyTest, Prefixing) {
+  auto table_key = LevelDbIndexEntryKey::KeyPrefix();
+
+  ASSERT_TRUE(absl::StartsWith(
+      LevelDbIndexEntryKey::Key(0, "user_id", "array_value_encoded",
+                                "directional_value_encoded", "document_id_99"),
+      table_key));
+
+  ASSERT_FALSE(absl::StartsWith(LevelDbIndexEntryKey::Key(0, "", "", "", ""),
+                                LevelDbIndexEntryKey::Key(1, "", "", "", "")));
+}
+
+struct IndexEntry {
```

Comment:
Not sure about this, but I would move this to the top of the file. You could also just store LevelDbIndexEntryKey directly.